modules/keymaps.nix

```diff
@@ -6,70 +6,119 @@
-          default = false;
-        };
-
-        unique = mkOption {
-          type = types.bool;
-          description = "Whether to fail if the map is already defined. Equivalent to adding <unique> to a map.";
-          default = false;
-        };
-
-        noremap = mkOption {
-          type = types.bool;
-          description = "Whether to use the 'noremap' variant of the command, ignoring any custom mappings on the defined action. It is highly advised to keep this on, which is the default.";
-          default = true;
-        };
-
-        remap = mkOption {
-          type = types.bool;
-          description = "Make the mapping recursive. Inverses \"noremap\"";
-          default = false;
-        };
+  # These are the configuration options that change the behavior of each mapping.
+  mapConfigOptions = {
+    silent =
+      helpers.defaultNullOpts.mkBool false
+      "Whether this mapping should be silent. Equivalent to adding <silent> to a map.";
+
+    nowait =
+      helpers.defaultNullOpts.mkBool false
+      "Whether to wait for extra input on ambiguous mappings. Equivalent to adding <nowait> to a map.";
+
+    script =
+      helpers.defaultNullOpts.mkBool false
+      "Equivalent to adding <script> to a map.";
+
+    expr =
+      helpers.defaultNullOpts.mkBool false
+      "Means that the action is actually an expression. Equivalent to adding <expr> to a map.";
+
+    unique =
+      helpers.defaultNullOpts.mkBool false
+      "Whether to fail if the map is already defined. Equivalent to adding <unique> to a map.";
+
+    noremap =
+      helpers.defaultNullOpts.mkBool true
+      "Whether to use the 'noremap' variant of the command, ignoring any custom mappings on the defined action. It is highly advised to keep this on, which is the default.";
+
+    remap =
+      helpers.defaultNullOpts.mkBool false
+      "Make the mapping recursive. Inverses \"noremap\"";
+
+    desc =
+      helpers.mkNullOrOption types.str
+      "A textual description of this keybind, to be shown in which-key, if you have it.";
+  };
 
-        action = mkOption {
-          type = types.str;
-          description = "The action to execute.";
+  # Generates maps for a lua config
+  genMaps = mode: maps: let
+    /*
+    Take a user-defined action (string or attrs) and return the following attribute set:
+    {
+      action = (string) the actual action to map to this key
+      config = (attrs) the configuration options for this mapping (noremap, silent...)
+    }
+
+    - If the action is a string:
+    {
+      action = action;
+      config = {};
+    }
+
+    - If the action is an attrs:
+    {
+      action = action;
+      config = {
+        inherit (action) <values of the config options that have been explicitly set by the user>
+      };
+    }
+    */
+    normalizeAction = action:
+      if isString action
+      # Case 1: action is a string
+      then {
+        inherit action;
+        config = helpers.emptyTable;
```

Comment:
Why do we want an emptyTable ? won't a simple {} not cause any problem as it will pickup the upstream defaults ?